tests/unittests/sources/test_cloudcix.py

```diff
@@ -0,0 +1,292 @@
+        assert self.datasource.metadata == METADATA
+        assert self.datasource.userdata_raw == USERDATA
+
+    @responses.activate
+    def test_failing_imds_endpoints(self):
+        # Make request before imds is set up
+        pytest.raises(
+            sources.InvalidMetaDataException,
+            self.datasource.crawl_metadata_service,
+        )
+
+        # Make imds respond to healthcheck
+        base_url = ds_mod.METADATA_URLS[0]
+        responses.add_callback(
+            responses.GET,
+            base_url,
+            callback=MockImds.base_response,
+        )
+
+        pytest.raises(
+            sources.InvalidMetaDataException,
+            self.datasource.crawl_metadata_service,
+        )
+
+        # Make imds serve metadata
+        version = ds_mod.METADATA_VERSION
+        responses.add_callback(
+            responses.GET,
+            uh.combine_url(base_url, f"v{version}", "metadata"),
+            callback=MockImds.metadata_response,
+        )
+
+        pytest.raises(
+            sources.InvalidMetaDataException,
+            self.datasource.crawl_metadata_service,
+        )
+
+        # Make imds serve userdata
+        responses.add_callback(
+            responses.GET,
+            uh.combine_url(base_url, f"v{version}", "userdata"),
+            callback=MockImds.userdata_response,
+        )
+
+        data = self.datasource.crawl_metadata_service()
+        assert data != dict()
+
+    @responses.activate
+    def test_read_malformed_metadata(self):
+        def bad_response(response):
+            return 200, response.headers, json.dumps(METADATA)[:-2]
+
+        version = ds_mod.METADATA_VERSION
+        base_url = ds_mod.METADATA_URLS[0]
+        versioned_url = uh.combine_url(base_url, f"v{version}")
+
+        # Malformed metadata
+        responses.add_callback(
+            responses.GET,
+            uh.combine_url(versioned_url, "metadata"),
+            callback=bad_response,
+        )
+        responses.add_callback(
+            responses.GET,
+            uh.combine_url(versioned_url, "userdata"),
+            callback=MockImds.userdata_response,
+        )
+
+        pytest.raises(
+            InvalidMetaDataException,
+            ds_mod.read_metadata,
+            versioned_url,
+            self.datasource.get_url_params(),
+        )
+
+    @responses.activate
+    def test_bad_response_code(self):
+        def bad_response(response):
+            return 404, response.headers, ""
+
+        version = ds_mod.METADATA_VERSION
+        base_url = ds_mod.METADATA_URLS[0]
+        versioned_url = uh.combine_url(base_url, f"v{version}")
+
+        responses.add_callback(
+            responses.GET,
+            uh.combine_url(versioned_url, "metadata"),
+            callback=bad_response,
+        )
+
+        pytest.raises(
+            InvalidMetaDataException,
+            ds_mod.read_metadata,
+            versioned_url,
+            self.datasource.get_url_params(),
+        )
```

Comment:
Let's specialize this assert to check for the expected error message
```suggestion
        with pytest.raises(
            InvalidMetaDataException,
            match="Invalid JSON at http://169.254.169.254/v1/metadata",
        ):
            ds_mod.read_metadata(
                versioned_url, self.datasource.get_url_params()
            )
```

Reply:
Ok, added that one.